### PR TITLE
crimson/net: add support for ms_learn_addr_from_peer.

### PR DESCRIPTION
**Pull request by rzarzynski**
An experimental, untested stuff which requires changes in the Seastar's network API (see https://github.com/rzarzynski/seastar/commit/e879050bccb9e1bfb59fa5882b4f703704dce3c2).

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/net/SocketMessenger.cc [hunk 1]
```diff
@@ -268,14 +268,27 @@ seastar::future<> SocketMessenger::learned_addr(const entity_addr_t &peer_addr_f
...
+                    // XXX: a syscall is here
+                    conn.get_local_address());
+      addr = peer_addr_for_me;
+    } else {
+      const auto local_addr_for_me = conn.get_local_address();
+      logger().info("{} socket to {} says I am {} (peer says {})",
+                    conn, conn.get_peer_socket_addr(),
+                    peer_addr_for_me);
+      addr.set_sockaddr(&local_addr_for_me.as_posix_sockaddr());
+    }
```
Comment (cyx1231st) [inline src/crimson/net/SocketMessenger.cc:285]: Encapsulate this to be shared below?

File: src/crimson/net/SocketMessenger.cc [hunk 2]
```diff
@@ -268,14 +268,27 @@ seastar::future<> SocketMessenger::learned_addr(const entity_addr_t &peer_addr_f
...
+      logger().info("{} peer {} says I am {} (socket says {})",
+                    conn, conn.get_peer_socket_addr(), peer_addr_for_me,
+                    // XXX: a syscall is here
+                    conn.get_local_address());
+      addr = peer_addr_for_me;
+    } else {
+      const auto local_addr_for_me = conn.get_local_address();
+      logger().info("{} socket to {} says I am {} (peer says {})",
+                    conn, conn.get_peer_socket_addr(),
+                    peer_addr_for_me);
```
Comment (cyx1231st) [inline src/crimson/net/SocketMessenger.cc:283]: Missing a parameter.

File: src/crimson/net/SocketMessenger.cc [hunk 3]
```diff
@@ -293,14 +307,27 @@ seastar::future<> SocketMessenger::learned_addr(const entity_addr_t &peer_addr_f
...
+        logger().info("{} peer {} says I am {} (socket says {})",
+                      conn, conn.get_peer_socket_addr(), peer_addr_for_me,
+                      // XXX: a syscall is here
+                      conn.get_local_address());
+        addr = peer_addr_for_me;
+      } else {
+        const auto local_addr_for_me = conn.get_local_address();
+        logger().info("{} socket to {} says I am {} (peer says {})",
+                      conn, conn.get_peer_socket_addr(),
+                      peer_addr_for_me);
```
Comment (cyx1231st) [inline src/crimson/net/SocketMessenger.cc:322]: Missing a parameter here, too.

File: src/crimson/net/SocketMessenger.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/net/SocketMessenger.cc]: Makes sense! Encapsulated.

File: src/crimson/net/SocketMessenger.cc (hunk 2 above)
Comment (rzarzynski) [reply src/crimson/net/SocketMessenger.cc]: Fixed.

File: src/crimson/net/SocketMessenger.cc (hunk 3 above)
Comment (rzarzynski) [reply src/crimson/net/SocketMessenger.cc]: Altogether with dissecting into a dedicated function.

**Comment by cyx1231st** [comment]
>> I suspect lossless connection would be broken in this scenario, for example two OSDs are both behind their own NAT. Incorrect peer_addr would be impossible for them to recover the connection.

> in the Rook arrangement OSDs are actually configured with the unusual ms_learn_addr_from_peer=false. OSDs are, clients are not.

@rzarzynski sorry I think I'm still confused.

Consider a scenario where both OSD.a and OSD.b are behind their own NAT:
* OSD.a binds to a local address-port Ai, externally remapped to Ae;
* OSD.b binds to a local address-port Bi, externally remapped to Be;

If OSDs are configured with `ms_learn_addr_from_peer=false`, they will not set their external addrs seen by peers during  `HelloFrame` exchange, instead they respect their local addresses, that is:
* OSD.a set to Ai;
* OSD.b set to Bi;

So, when OSDs try to connect each other, for example, OSD.a connects to OSD.b. Then in OSD.b's `server_connect()`, it will set its `conn.peer_addr` to Ai from OSD.a's `client_ident.addrs.front()`.

After the connection is established and then broken, if OSD.b tries to reconnect to OSD.a, it will use `conn.peer_addr` Ai. But that should be result in failure because OSD.b can only reach OSD.a through Ae, right?

Is my above understanding correct?

**Comment by cyx1231st** [comment]
>> I suspect lossless connection would be broken in this scenario, for example two OSDs are both behind their own NAT. Incorrect peer_addr would be impossible for them to recover the connection.

> in the Rook arrangement OSDs are actually configured with the unusual ms_learn_addr_from_peer=false. OSDs are, clients are not.

@rzarzynski sorry I think I'm still confused.

Consider a scenario where both OSD.a and OSD.b are behind their own NAT:
* OSD.a binds to a local address-port Ai, externally remapped to Ae;
* OSD.b binds to a local address-port Bi, externally remapped to Be;

If OSDs are configured with `ms_learn_addr_from_peer=false`, they will not set their external addrs seen by peers during  `HelloFrame` exchange, instead they respect their local addresses, that is:
* OSD.a set to Ai;
* OSD.b set to Bi;

So, when OSDs try to connect each other, for example, OSD.a connects to OSD.b. Then in OSD.b's `server_connect()`, it will set its `conn.peer_addr` to Ai from OSD.a's `client_ident.addrs.front()`.

After the connection is established and then broken, if OSD.b tries to reconnect to OSD.a, it will use `conn.peer_addr` Ai. But that should be result in failure because OSD.b can only reach OSD.a through Ae, right?

Is my above understanding correct?

**Comment by rzarzynski** [comment]
>  Consider a scenario where both OSD.a and OSD.b are behind their own NAT:
>
>   * OSD.a binds to a local address-port Ai, externally remapped to Ae;
>   * OSD.b binds to a local address-port Bi, externally remapped to Be;
> (...)
> After the connection is established and then broken, if OSD.b tries to reconnect to OSD.a, it will use `conn.peer_addr` Ai. But that should be result in failure because OSD.b can only reach OSD.a through Ae, right?

@cyx1231st: there is a kludge. Both Ai and Bi are **routable in boundaries of an entire k8s cluster**. Initially, I also thought they are not and the NAT is supposed to overcome that but I was wrong. **Ai and Bi can be used to for inter-cluster traffic even when are they assigned to pods on different nodes**.

The problem NAT is supposed to tackle in Kubernetes is basically that pod addresses are ephemeral -- a network arrangement can easily can change (due to e.g. pod restart, scaling number of them etc.). This isn't a big deal for an OSD who wants to talk to another OSD thankfully to monitors and `OSDMap`. However, finding a monitor is a completely different story. Kubernetes brings here the concepts of _Service_ and persistent (as I understand it) _Cluster IP_ Rook uses to make `mon_host` stable. Single _Cluster IP_ can represent multiple pods and load-balancing between them is another reason for headache. There are multiple modes for doing that but the net result is that SNAT **can** (but not always must) happen, and thus source IP can change when talking with a service. Please take a look on [this blog post](https://www.asykim.com/blog/deep-dive-into-kubernetes-external-traffic-policies) describing quirky details of some traffic policies.

When it comes to the address learning and `ms_learn_addr_from_peer`: I bet an OSD / a client (always?) learns their messengers' IPs from monitors as a monitor is contacted at the very beginning. To overcome the learning snated Ae and Be OSDs are ignoring the entries in `HelloFrame` but clients, as they keep the default `true` for `ms_learn_addr_from_peer`, are not. This goes (entirely?) unpushined when talking to classical OSDs as they don't have the (very sound, BTW) check crimson had.

**Comment by cyx1231st** [comment]
@rzarzynski thanks for the detailed description and the background knowledge provided! I think I've got the point why the current address exchange and learning work like this, as the monitor-OSD, OSD-OSD and client-OSD are connected under different network configurations in the k8s scenario, and there are some workarounds that looks tricky right now until we can figure out a cleaner solution to integrate with k8s.

**Comment by rzarzynski** [comment]
> there are some workarounds that looks tricky right now until we can figure out a cleaner solution to integrate with k8s.

Fully agreed, I share the feeling that _it works by an accident_. Some time ago I (optimistically) thought the problem of not having `ms_learn_addr_from_peer` set to `false` is restricted to `ceph-mgr` only. Experimenting proved it wrong – clients are using the default as well. This deceased my initial intention to fill a bug report against Rook as I doubt there is a control over all RADOS clients in k8s cluster.

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
@liu-chunmei want to take a look?

**Review by cyx1231st — APPROVED** [review]
LGTM

**Comment by rzarzynski** [comment]
The `local_address()` change has been merged with Seastar. Unfortunately, incorporating fresh `master` of Seastar requires more than just bumping a submodule up, so I will proceed the upgrade as a separated PR.

For this PR placing the current `local_address()` patch on top of our fork's current `master`.

**Review by liu-chunmei — APPROVED** [review]
LGTM.  This PR is not fix my problem. seems I meet another one.

**Comment by rzarzynski** [comment]
Rebased to fresh `master` to make this PR free of modifying the `seastar` submodule (the Seastar part is already in).

**Comment by cyx1231st** [comment]
jenkins retest this please

**Pull request merged**